Attestation/ek-secret.md

```diff
@@ -0,0 +1,173 @@
+    return ekctx
+
+def eksession(ectx):
+    session = ectx.StartAuthSession(
+        ESYS_TR.NONE,
+        ESYS_TR.NONE,
+        None,
+        TPM2_SE.POLICY,
+        TPMT_SYM_DEF(algorithm=TPM2_ALG.NULL),
+        TPM2_ALG.SHA256,
+    )
+
+    ectx.PolicySecret(
+        ESYS_TR.RH_ENDORSEMENT,
+        session,
+        TPM2B_NONCE()._cdata,
+        TPM2B_DIGEST()._cdata,
+        TPM2B_NONCE()._cdata,
+        0,
+        session1=ESYS_TR.PASSWORD,
+    )
+    
+    return session
+
+def main(ekpath, publicpath, sensitivepath, credpath, secretpath):
+    # time to setup a ESAPI context, we will use the default TCTI for the system
+    ectx = ESAPI()
+
+    # Time to load the EK context, by using tpm2_createek there is no reason the implement the whole setup in this example code
+    with open(ekpath, 'rb') as ekf:
+        ekb = ekf.read()
+    ekctx = unmarshal_tools_context(ekb)
+    ekhandle = ectx.ContextLoad(ekctx)
+
+    # now lets setup the standard EK policy session
+    session = eksession(ectx)
+    
+    # Now we should read, unmarshal and load the temporary key pair
+    with open(publicpath, 'rb') as pubf:
+        pubb = pubf.read()
+    public, _ = TPM2B_PUBLIC.Unmarshal(pubb)
+    with open(sensitivepath, 'rb') as sensf:
+        sensb = sensf.read()
+    sensitive, _ = TPM2B_SENSITIVE.Unmarshal(sensb)
+    print(sensitive.sensitiveArea.authValue.size, public.publicArea.authPolicy.size)
+    # We will load it under the NULL hierarchy as that is the only hierarchy allowing both the public and private part to be loaded for external keys
+    handle = ectx.LoadExternal(sensitive, public, ESYS_TR.RH_NULL)
+    
+    
+    # Time to read and unmarshal the credential and secret for ActivateCredential
+    with open(credpath, 'rb') as credf:
+        credb = credf.read()
+    credblob, _ = TPM2B_ID_OBJECT.Unmarshal(credb)
+    with open(secretpath, 'rb') as secretf:
+        secretb = secretf.read()
+    secret, _ = TPM2B_ENCRYPTED_SECRET.Unmarshal(secretb)
+    
+    # Well, now there is nothing left but calling ActivateCredential and getting our secret on the remove system!
+    oursecret = ectx.ActivateCredential(handle, ekhandle, credblob, secret, session2=session)
+
+    print(f"we got the secret: {bytes(oursecret)}")
+
+if __name__ == '__main__':
+    if len(sys.argv) < 6:
+        sys.stderr.write(f"usage: {sys.argv[0]} ek-ctx temp-public temp-sensitive credblob secret\n")
+        exit(1)
+    main(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])
+```
+
+Generate the EK context with `tpm2_createek -c ek.ctx`
+The arguments are:
+ek-ctx: the context generated by tpm2_createek
+temp-public: the temp-public output from the local system script
+temp-sensitive: the temp-sensitive output from the local system script
+credblob: the credblob output from the local system script
+secret: the secret output from the local system script
+
+## Issues
+there is no protection against replay attacks in this example
```

Comment:
Replay protection can be had by: a) adding a timestamp to the plaintext message, b) adding a replace cache on the remote system.

Another issue is that the sender is not authenticated.  That can be added by adding a signature.